### ISSUE TITLE
[Bug] A matcher with multiple eval() doesn't work

### ISSUE THREAD

**Issue by tvanriper** [casbin/casbin#619]
**Want to prioritize this issue? Try:**

[![issuehunt-to-marktext](https://github.com/BoostIO/issuehunt-materials/raw/master/v1/issuehunt-button-v1.svg)](https://issuehunt.io/r/casbin/casbin)

------

**Describe the bug**
A matcher with multiple eval() statements only calls the first eval(), caching the results.

**To Reproduce**
Compile this:

```go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"

	"github.com/casbin/casbin/v2"
	"github.com/casbin/casbin/v2/model"
)

// Structures used for the subject and object.

type OrgPerm struct {
	Name string
	Role string
}

type User struct {
	ID   int
	Name string
	Perm OrgPerm
}

type TableData struct {
	Columns []string `json:"columns"`
	Table   string   `json:"table"`
}

type GuiData struct {
	Role string `json:"role"`
}

// Policies and model for 'double eval' approach.

func getDoublePolicies() [][]string {
	return [][]string{
		{"strCmp(r.sub.Perm.Role, \"admin\") == 0", "contains(r.obj.Columns, \"email\")", "write"},
		{"strCmp(r.sub.Perm.Role, \"admin\") == 0", "contains(r.obj.Columns, \"email\")", "read"},
		{"strCmp(r.sub.Perm.Role, \"instructor\") == 0", "contains(r.obj.Columns, \"Jojo\")", "read"},
	}
}

func getDoubleModel() (*model.Model, error) {
	result := model.NewModel()
	err := result.LoadModelFromText(`
[request_definition]
r = sub, obj, act
[policy_definition]
p = sub_rule, obj_test, act
[policy_effect]
e = some(where (p.eft == allow))
[matchers]
m = eval(p.sub_rule) && eval(p.obj_test) && strCmp(r.act, p.act) == 0
`)
	return &result, err
}

// Policies and model for 'single eval' approach.

func getSinglePolicies() [][]string {
	return [][]string{
		{"strCmp(r.sub.Perm.Role, \"admin\") == 0", "email", "write"},
		{"strCmp(r.sub.Perm.Role, \"admin\") == 0", "email", "read"},
		{"strCmp(r.sub.Perm.Role, \"instructor\") == 0", "Jojo", "read"},
	}
}

func getSingleModel() (*model.Model, error) {
	result := model.NewModel()
	err := result.LoadModelFromText(`
[request_definition]
r = sub, obj, act
[policy_definition]
p = sub_rule, obj_test, act
[policy_effect]
e = some(where (p.eft == allow))
[matchers]
m = eval(p.sub_rule) && contains(r.obj.Columns, p.obj_test) && strCmp(r.act, p.act) == 0
`)
	return &result, err
}

// Functions we'll add to the evaluator.

func contains(args ...interface{}) (interface{}, error) {
	fmt.Printf("contains( %#v ) = ", args)
	result := false
	arr, ok := args[0].([]string)
	if !ok {
		fmt.Println("Bad first arg")
		return result, errors.New("contains: 1st arg is not a string array")
	}
	test, ok := args[1].(string)
	if !ok {
		fmt.Println("Bad second arg")
		return result, errors.New("contains: 2nd arg is not a string")
	}
	for _, item := range arr {
		if item == test {
			result = true
			break
		}
	}
	fmt.Printf("%#v\n", result)
	return result, nil
}

func strCmp(args ...interface{}) (interface{}, error) {
	fmt.Printf("strCmp( %#v ) returns: ", args)
	result := -1
	left, ok := args[0].(string)
	if !ok {
		fmt.Println("Bad first arg")
		return result, errors.New("strCmp: 1st arg is not a string")
	}
	right, ok := args[1].(string)
	if !ok {
		fmt.Println("Bad second arg")
		return result, errors.New("strCmp: 2nd arg is not a string")
	}
	result = strings.Compare(left, right)

	fmt.Printf("%#v\n", result)

	return (float64)(result), nil
}

// Enforce provides some nice logging around the enforcer as we use it.

func Enforce(e *casbin.Enforcer, sub User, obj interface{}, act string) {
	fmt.Println("------------------Enforce---------------------")
	ok, why, err := e.EnforceEx(sub, obj, act)
	fmt.Println("----------------------------------------------")
	if err != nil {
		fmt.Printf("Enforce failed: %s\n", err)
	}
	if ok {
		fmt.Printf("User %s permitted to access %s for %#v\n", sub.Name, obj, act)
		fmt.Println("Here's why:")
		for _, item := range why {
			fmt.Println(item)
		}
	} else {
		fmt.Printf("User %s is not allowed to access %s for %#v\n", sub.Name, obj, act)
	}
}

// Testing with the 'double eval' version of the policy and model.

func doubleEval() {
	e, err := casbin.NewEnforcer()
	if err != nil {
		fmt.Printf("Failed to acquire enforcer: %s\n", err)
		os.Exit(1)
	}
	e.EnableLog(true)
	model, err := getDoubleModel()
	if err != nil {
		fmt.Printf("Failed to create model: %s\n", err)
		os.Exit(1)
	}
	e.SetModel(*model)
	e.GetModel().PrintModel()
	_, err = e.AddPolicies(getDoublePolicies())
	if err != nil {
		fmt.Printf("Failed to add policies: %s\n", err)
		os.Exit(1)
	}
	e.AddFunction("strCmp", strCmp)
	e.AddFunction("contains", contains)

	sub := User{
		ID:   1,
		Name: "bob",
		Perm: OrgPerm{
			Name: "CYBRScore",
			Role: "admin",
		},
	}
	obj := TableData{
		Columns: []string{
			"email",
			"blubber",
		},
		Table: "table",
	}
	act := "read"

	fmt.Println("=========================================================")
	fmt.Println("Testing with 'admin' role (should pass).")
	Enforce(e, sub, obj, act)

	fmt.Println("=========================================================")
	fmt.Println("Testing with 'instructor' role (should fail).")
	sub.Perm.Role = "instructor"
	Enforce(e, sub, obj, act)

	fmt.Println("=========================================================")
	fmt.Println("Testing with 'instructor' role, columns with 'Jojo' (should pass).")
	obj.Columns = []string{
		"yellow",
		"blue",
		"Jojo",
	}
	Enforce(e, sub, obj, act)

	act = "write"
	fmt.Println("=========================================================")
	fmt.Println("Testing with 'admin' role write access after adding 'email' to columns (should pass).")
	sub.Perm.Role = "admin"
	obj.Columns = append(obj.Columns, "email")
	Enforce(e, sub, obj, act)
}

// Testing with the 'single eval' version of the policy and model.

func singleEval() {
	e, err := casbin.NewEnforcer()
	if err != nil {
		fmt.Printf("Failed to acquire enforcer: %s\n", err)
		os.Exit(1)
	}
	e.EnableLog(true)
	model, err := getSingleModel()
	if err != nil {
		fmt.Printf("Failed to create model: %s\n", err)
		os.Exit(1)
	}
	e.SetModel(*model)
	e.GetModel().PrintModel()
	_, err = e.AddPolicies(getSinglePolicies())
	if err != nil {
		fmt.Printf("Failed to add policies: %s\n", err)
		os.Exit(1)
	}
	e.AddFunction("strCmp", strCmp)
	e.AddFunction("contains", contains)

	sub := User{
		ID:   1,
		Name: "bob",
		Perm: OrgPerm{
			Name: "CYBRScore",
			Role: "admin",
		},
	}
	obj := TableData{
		Columns: []string{
			"email",
			"blubber",
		},
		Table: "table",
	}
	act := "read"

	fmt.Println("=========================================================")
	fmt.Println("Testing with 'admin' role (should pass).")
	Enforce(e, sub, obj, act)

	fmt.Println("=========================================================")
	fmt.Println("Testing with 'instructor' role (should fail).")
	sub.Perm.Role = "instructor"
	Enforce(e, sub, obj, act)

	fmt.Println("=========================================================")
	fmt.Println("Testing with 'instructor' role, columns with 'Jojo' (should pass).")
	obj.Columns = []string{
		"yellow",
		"blue",
		"Jojo",
	}
	Enforce(e, sub, obj, act)

	act = "write"
	fmt.Println("=========================================================")
	fmt.Println("Testing with 'admin' role write access after adding 'email' to columns (should pass).")
	sub.Perm.Role = "admin"
	obj.Columns = append(obj.Columns, "email")
	Enforce(e, sub, obj, act)
}

func main() {
	fmt.Println("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=")
	fmt.Println("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=singleEval()-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=")
	singleEval()
	fmt.Println("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=")
	fmt.Println("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=doubleEval()-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=")
	doubleEval()
}
```
... then run it.

Notice in the logging how the singleEval() shows the 'contains()' function gets called whenever the first match succeeds, but the doubleEval() never shows logging for 'contains()'.

Also notice that the third evaluated request fails for singleEval, but passes for doubleEval.

**Expected behavior**

Both singleEval and doubleEval should have the same logging.

**Comment by nodece** [comment]
@tvanriper I submitted a PR: https://github.com/casbin/casbin/pull/620. could you review this PR?

**Comment by tvanriper** [comment]
> @tvanriper I submitted a PR: #620. could you review this PR?

It looks okay, but I lack familiarity with this code base.  I would defer to those who know the code base better.

**Comment by hsluoyz** [comment]
@tvanriper released v2.14.2: https://github.com/casbin/casbin/releases/tag/v2.14.2
